### DISCUSSION TITLE
Why does ";;" in the script remove redundant buffers?

### DISCUSSION THREAD

**Discussion by gatk555** [YosysHQ/yosys#3901] [Q&A] [answered]
Please excuse any ignorance; I am new at this.  I am using 0.32 with scripts like this (no code blocks, as they seem not to work today):

read_verilog minimal.v
read_verilog -lib prim_cells.v

proc; memory; techmap

dfflibmap -liberty prim_cells.lib
abc -liberty prim_cells.lib;;

write_verilog minimal_syn.v
write_spice -neg d_low -pos d_high -buf BUF minimal.subckt

With this input:

module my_nand(a, b, c);
   input wire a, b;
   output wire c;
   assign c = ~(a & b);
endmodule // my_and

I get the expected:

.SUBCKT my_nand a b c
X0 a b 1 NAND
.ENDS my_nand

However, if I remove a semicolon from the "abc" line in the script the output becomes:

.SUBCKT my_nand a b c
X0 a b 1 NAND
X1 a 2 BUF
X2 b 3 BUF
X3 1 c BUF
.ENDS my_nand

If I compare the logging output, it gets weirder.  The only significant difference is that with the semicolons:

Removed 0 unused cells and 1 unused wires.

appears in the proc pass.  Can anyone explain this?

**Comment by whitequark** [comment, marked as answer]
`;;` is equivalent to `clean` (and `;;;` is equivalent to `clean -purge`). See `help opt_clean`.

**Reply by gatk555** [reply]
So undocumented magic?

Edit: actually not undocumented as "help clean" describes it.  The documentation is just hidden behind a door labelled "Beware of the leopard".

Thanks.

**Reply by whitequark** [reply]
It's documented in several places; aside from the one I gave you, [here](https://yosyshq.readthedocs.io/projects/yosys/en/latest/appendix/APPNOTE_011_Design_Investigation.html) and [here](https://yosyshq.readthedocs.io/projects/yosys/en/latest/cmd/clean.html).
